### ISSUE TITLE
Running Nomad in docker container

### ISSUE THREAD

**Issue by agarwalpradeep123** [hashicorp/nomad#542]
We are running nomad server and nomad client in docker containers.

Nomad Server configuration:
log_level = "DEBUG"
bind_addr = "0.0.0.0"
data_dir = "/tmp/server"
advertise {
    rpc = "10.245.1.2:4647"
    serf = "10.245.1.2:4648"
}
server {
    enabled = true
    bootstrap_expect = 1
}

Nomad Client configuration:
log_level = "DEBUG"
bind_addr = "0.0.0.0"
data_dir = "/tmp/client"
client {
    enabled = true
    servers = ["10.245.1.2:4647"]
}

Nomad Client logs:
==> Starting Nomad agent...
2015/12/07 16:50:32 [DEBUG] fingerprint.env_aws: Error querying AWS Metadata URL, skipping
==> Nomad agent configuration:

```
             Atlas: <disabled>
            Client: true
         Log Level: DEBUG
            Region: global (DC: dc1)
            Server: false
```

==> Nomad agent started! Log data will stream in below:

```
2015/12/07 16:50:30 [INFO] client: using state directory /tmp/client/client
2015/12/07 16:50:30 [INFO] client: using alloc directory /tmp/client/alloc
2015/12/07 16:50:30 [DEBUG] client: periodically fingerprinting consul at duration 15s
2015/12/07 16:50:34 [DEBUG] fingerprint.env_gce: Error querying GCE Metadata URL, skipping
2015/12/07 16:50:34 [DEBUG] fingerprint.network: Detected interface eth0  with IP 172.17.0.4 during fingerprinting
2015/12/07 16:50:34 [DEBUG] client: applied fingerprints [arch cpu host memory network storage]
2015/12/07 16:50:34 [DEBUG] driver.docker: using client connection initialized from environment
2015/12/07 16:50:34 [DEBUG] driver.docker: privileged containers are disabled
2015/12/07 16:50:34 [DEBUG] client: available drivers [docker exec]
2015/12/07 16:50:34 [DEBUG] client: node registration complete
2015/12/07 16:50:34 [DEBUG] client: updated allocations at index 1 (0 allocs)
2015/12/07 16:50:34 [DEBUG] client: allocs: (added 0) (removed 0) (updated 0) (ignore 0)
2015/12/07 16:50:42 [DEBUG] client: state updated to ready
```

When I submitted a job to nomad server, allocation seems to fail
Output of curl localhost:4646/v1/allocations,
[{
    "ID": "ce174c9d-874a-49b8-6643-1a9aea38dbdc",
    "EvalID": "cf20c64f-9748-7c14-e37d-59f8e6319140",
    "Name": "sample.webs[0]",
    "NodeID": "",
    "JobID": "sampleID",
    "TaskGroup": "webs",
    "DesiredStatus": "failed",
    "DesiredDescription": "failed to find a node for placement",
    "ClientStatus": "failed",
    "ClientDescription": "",
    "TaskStates": {
        "frontend": {
            "State": "dead",
            "Events": null
        }
    },
    "CreateIndex": 141,
    "ModifyIndex": 141
}]

I think this is caused by fingerprint.network being detected as 172.17.0.4(in client logs). This is a internal IP address of docker container running nomad client. But I could not find any configuration which can be used to set this IP address to something which is reachable from other hosts.

Could you please help me on this?

**Comment by diptanu** [comment]
@agarwalpradeep123 Please look at the Client Specific Configuration on the documentation page - https://nomadproject.io/docs/agent/config.html

It is possible to force the fingerprinter to use a specific network interface by specifying the ```network_interface``` option in the configuration file.

**Comment by agarwalpradeep123** [comment]
network_interface option can be used to specify interfaces which are available inside docker container, but I want to specify host IP on which container is running. Still I had tried to use network_interface option by setting it to "10.245.1.10" but it fails saying interface not found.

To give more information,
Output of curl localhost:4646/v1/node/e94bf03d-5e44-5401-0c5d-c10bb5d2e7ba
{"ID":"e94bf03d-5e44-5401-0c5d-c10bb5d2e7ba","Datacenter":"dc1","Name":"d30abc45d0b8","Attributes":{"arch":"amd64","cpu.frequency":"2893.426000","cpu.modelname":"Intel(R) Core(TM) i7-3520M CPU @ 2.90GHz","cpu.numcores":"1","cpu.totalcompute":"2893.426000","driver.docker":"1","driver.docker.version":"1.9.0","driver.exec":"1","hostname":"d30abc45d0b8","kernel.name":"linux","kernel.version":"4.2.5-201.fc22.x86_64","memory.totalbytes":"1041670144","network.ip-address":"172.17.0.4","os.name":"fedora","os.version":"22","storage.bytesfree":"99700006912","storage.bytestotal":"105555197952","storage.volume":"/dev/mapper/docker-8:3-134574965-d30abc45d0b8cb712549c6216131ce122719af4d18bc445aa47fbbf7451bfb6d"},"Resources":{"CPU":2893,"MemoryMB":993,"DiskMB":95081,"IOPS":0,"Networks":[{"Device":"eth0","CIDR":"172.17.0.4/32","IP":"172.17.0.4","MBits":10000,"ReservedPorts":null,"DynamicPorts":null}]},"Reserved":null,"Links":{},"Meta":{},"NodeClass":"","Drain":false,"Status":"down","StatusDescription":"","CreateIndex":65,"ModifyIndex":829}

As you can see from above output that nomad client IP which nomad server knows is "172.17.0.4". This IP address is not reachable by server and is internal IP of nomad client container.

How can I solve this?

**Comment by diptanu** [comment]
@agarwalpradeep123 Nomad expects the network_interface option to be the network interface name and not the ip address.

**Comment by agarwalpradeep123** [comment]
I am not able to get it to work yet. Not sure if you are getting my problem.

As per reply from nomad server for /v1/node/<node_id>, it seems that nomad server thinks nomad client IP address is "172.17.0.4". Server might be using this IP address to communicate job/task to be executed and it fails to contact node. I do not see anything in client logs. I am not sure if there is any other cause of the failure.

Please let me know if you need more information.

**Comment by diptanu** [comment]
@agarwalpradeep123 So you're trying to make the Nomad client use an interface which doesn't exist in the container. I am guessing your current setup is using the Docker bridge networking mode? 

The only interface the client is seeing is the network interface inside the container. So if you want to use a host network interface make sure you're running the container with Nomad client with --net=host. That will make the nomad client see all the network interfaces of the host.

Hope that helps.

**Comment by agarwalpradeep123** [comment]
I tried with --net=host and setting network_interface="eth1" as default it uses is "eth0". With this it worked.

But I cannot configure network_interface statically, which can be different on different machines/installations. Host IP address is known before hand, it would be good if there is way to supply host IP to nomad client which will be used to start docker container for applications. Do you see this as an enhancement request?

**Comment by cdrage** [comment]
@agarwalpradeep123 

check out my blog post at http://charliedrage.com/nomad it addresses your issue as I ran into the same thing before.

**Comment by agarwalpradeep123** [comment]
@cdrage Thanks for reply. Your blog suggests same thing which is start docker container with "--net=host" and specify network_interface while starting nomad client docker container. But I cannot configure network_interface statically in my scripts, which can be different on different machines/installations. Host IP address is known before hand, it would be good if there is way to supply host IP to nomad client which will be used to start docker container for applications and hence host IP will be used for binding ports.

**Comment by diptanu** [comment]
@agarwalpradeep123 I would suggest using your provisioning script to find out the network interface name of the IP you want the Nomad agent to use and put it in the configuration file.

**Comment by agarwalpradeep123** [comment]
I don't think that is a good way of doing it. There will be different ways to get this information on different OS. I think it should be allowed to pass IP address which will be externally accessible.

**Comment by dcharbonnier** [comment]
I think we should be able to specify the address when the services check are registered the ip is the nomad container ip and consul can't perform the check.
for the consul registrator (http://gliderlabs.com/registrator/latest/user/run/) there is an --ip option

**Comment by Edenshaw** [comment]
I know this issue is closed, but in case you are facing the same problem, I discovered that the order of the commands is relevant in order to pick up the right interface.

If you issue the command:

`nomad agent -config=/etc/nomad/nomad-client.hcl -client 192.168.56.122 -network-interface eth1`

It won't work, it will still pick up eth0; but if you try using:

`nomad agent -config=/etc/nomad/nomad-client.hcl -network-interface eth1 -client 192.168.56.122`

Then it will work. Eureka! (This should be documented somewhere ¬¬)

PS: Remeber, there are two ways of setting the network_interface option: 

1. **network_interface**: It will work in the configuration of the client in a hcl file

2. **network-interface**: It will work only as a argument to the agent command of nomad, not in a configuration file. In the previous explanation I did, I used this option: as a command.

Although all the settings are done in the nomad client, I checked the server logs of nomad (in Debug mode) and there's also a fingerprint.network process (it was picking eth0 without the above command) so I also add the network-interface option to the server in order to be in the same page of the nomad client.
